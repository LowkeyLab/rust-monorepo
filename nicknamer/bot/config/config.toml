[nicknamer.reveal]
insult = "ya dingus"
role_to_mention = "Code Monkeys"
he_who_shall_not_be_named = 899501665365929985
