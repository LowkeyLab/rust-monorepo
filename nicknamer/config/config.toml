[nicknamer.reveal]
insult = "ya dingus"
role_to_mention = "Code Monkeys"