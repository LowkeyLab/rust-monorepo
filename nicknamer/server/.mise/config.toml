[env]
'_'.file = '.env'

[tools]
"cargo:bacon" = "latest"
