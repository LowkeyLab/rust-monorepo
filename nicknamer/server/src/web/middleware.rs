use axum::extract::Request;
use axum::http::{HeaderName, HeaderValue};
use axum::middleware::Next;
use axum::response::Response;

/// Middleware function that adds CORS headers to expose HTMX headers
pub async fn cors_expose_headers(request: Request, next: Next) -> Response {
    let mut response = next.run(request).await;

    // Add the Access-Control-Expose-Headers header
    response.headers_mut().insert(
        HeaderName::from_static("access-control-expose-headers"),
        HeaderValue::from_static("hx-retarget,hx-reswap"),
    );

    response
}

#[cfg(test)]
mod tests {
    use super::*;
    use axum::body::Body;
    use axum::http::{Request, StatusCode};
    use axum::{Router, response::Response};
    use tower::ServiceExt;

    #[tokio::test]
    async fn can_add_cors_expose_header() {
        let app = Router::new()
            .route("/test", axum::routing::get(|| async { "test response" }))
            .layer(axum::middleware::from_fn(cors_expose_headers));

        let response = app
            .oneshot(
                Request::builder()
                    .method("GET")
                    .uri("/test")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        let headers = response.headers();
        let expose_headers = headers.get("access-control-expose-headers");
        assert_eq!(
            expose_headers,
            Some(&axum::http::HeaderValue::from_static(
                "hx-retarget,hx-reswap"
            ))
        );
    }

    #[tokio::test]
    async fn can_preserve_existing_headers_when_adding_cors() {
        async fn handler_with_custom_header() -> Response<String> {
            let mut response = Response::new("test response".to_string());
            response.headers_mut().insert(
                "custom-header",
                axum::http::HeaderValue::from_static("custom-value"),
            );
            response
        }

        let app = Router::new()
            .route(
                "/test-with-headers",
                axum::routing::get(handler_with_custom_header),
            )
            .layer(axum::middleware::from_fn(cors_expose_headers));

        let response = app
            .oneshot(
                Request::builder()
                    .method("GET")
                    .uri("/test-with-headers")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        let headers = response.headers();

        // Check that our CORS header was added
        let expose_headers = headers.get("access-control-expose-headers");
        assert_eq!(
            expose_headers,
            Some(&axum::http::HeaderValue::from_static(
                "hx-retarget,hx-reswap"
            ))
        );

        // Check that existing headers are preserved
        let custom_header = headers.get("custom-header");
        assert_eq!(
            custom_header,
            Some(&axum::http::HeaderValue::from_static("custom-value"))
        );
    }
}
