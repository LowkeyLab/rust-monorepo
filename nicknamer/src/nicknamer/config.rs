#[allow(dead_code)]
pub const REVEAL_INSULT: &str = "ya dingus";
#[allow(dead_code)]
pub const CODE_MONKEYS_ROLE_NAME: &str = "Code Monkeys";
#[allow(dead_code)]
pub const URL_LENGTH_VIOLATION_FACTOR: u32 = 2;
#[allow(dead_code)]
pub const JAR_JAR_EMOJI_ID: u64 = 1061775549065855079;
#[allow(dead_code)]
pub const ZACH_USER_ID: u64 = 894692357457469471;
#[allow(dead_code)]
pub const JAR_JAR_COLOR_HEX: u32 = 0xD59D7E;
#[allow(dead_code)]
pub const HORRIFIED_JAR_JAR_PIC: &str =
    "https://cdn.mos.cms.futurecdn.net/RvLDChLaR37NWTEjvQm2pB-970-80.jpg.webp";
#[allow(dead_code)]
pub const HAPPY_JAR_JAR_PIC: &str =
    "https://static.wikia.nocookie.net/unanything/images/c/c7/Jar_Jar.jpg/revision/latest";
