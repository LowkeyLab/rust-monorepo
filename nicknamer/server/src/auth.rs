use askama::Template;
use axum::Router;
use axum::extract::{Extension, Form, Request, State};
use axum::http::{HeaderMap, HeaderName, HeaderValue};
use axum::middleware::{self, Next};
use axum::response::{Html, IntoResponse, Response};
use axum_extra::extract::CookieJar;
use jsonwebtoken::encode;
use std::sync::Arc;

use crate::config::Config;

/// Represents the currently authenticated user.
#[derive(Debug, Clone)]
pub struct CurrentUser {
    pub username: String,
}

impl CurrentUser {
    /// Creates a new CurrentUser instance.
    pub fn new(username: String) -> Self {
        Self { username }
    }
}

/// Authentication state containing admin credentials and JWT secret.
#[derive(Clone)]
pub struct AuthState {
    pub admin_username: String,
    pub admin_password: String,
    pub jwt_secret: String,
}

impl AuthState {
    /// Creates a new AuthState from the application config.
    pub fn from_config(config: &Config) -> Self {
        Self {
            admin_username: config.admin_username.clone(),
            admin_password: config.admin_password.clone(),
            jwt_secret: config.jwt_secret.clone(),
        }
    }
}

/// Creates a login router with authentication routes.
pub fn create_login_router() -> Router<Arc<AuthState>> {
    Router::new().route("/login", axum::routing::post(login_handler))
}

/// Creates a router with authentication middleware.
pub fn create_auth_router() -> Router<Arc<AuthState>> {
    Router::new().layer(middleware::from_fn_with_state(
        Arc::new(AuthState {
            admin_username: String::new(),
            admin_password: String::new(),
            jwt_secret: String::new(),
        }),
        auth_middleware,
    ))
}

/// Authentication middleware that checks for valid JWT tokens and sets CurrentUser extension.
pub async fn auth_middleware(
    State(state): State<Arc<AuthState>>,
    jar: CookieJar,
    mut request: Request,
    next: Next,
) -> Response {
    if let Some(token_cookie) = jar.get("auth_token") {
        if let Ok(claims) = decode_jwt(token_cookie.value(), &state.jwt_secret).await {
            let current_user = CurrentUser::new(claims.username);
            request.extensions_mut().insert(current_user);
        }
    }

    next.run(request).await
}

/// Represents the login request payload.
#[derive(serde::Deserialize, Debug)]
pub struct LoginRequest {
    pub username: String,
    pub password: String,
}

#[derive(serde::Serialize, serde::Deserialize, Debug)]
pub struct Claims {
    pub exp: usize,       // Expiry time of the token
    pub iat: usize,       // Issued at time of the token
    pub username: String, // Username of the authenticated user
}
/// Custom error type for authentication operations.
#[derive(Debug, thiserror::Error)]
pub enum AuthError {
    /// Represents an error during template rendering.
    /// The specific `askama::Error` is captured as the source of this error.
    #[error("Template rendering failed")]
    Template(#[from] askama::Error),
    /// Represents an error during JWT operations.
    /// The specific `jsonwebtoken::errors::Error` is captured as the source of this error.
    #[error("JWT operation failed")]
    JwtError,
}

impl axum::response::IntoResponse for AuthError {
    fn into_response(self) -> axum::response::Response {
        let user_facing_error_message =
            "An unexpected error occurred while processing your request. Please try again later.";
        (
            axum::http::StatusCode::INTERNAL_SERVER_ERROR,
            Html(format!(
                "<h1>Internal Server Error</h1><p>{}</p>",
                user_facing_error_message
            )),
        )
            .into_response()
    }
}

/// Handles the login request.
/// Checks submitted username and password against admin credentials.
/// If a user is already logged in, returns a success message.
pub async fn login_handler(
    State(state): State<Arc<AuthState>>,
    jar: CookieJar,
    current_user: Option<Extension<CurrentUser>>,
    Form(payload): Form<LoginRequest>,
) -> Result<(CookieJar, Response), AuthError> {
    // Check if user is already logged in
    if let Some(Extension(user)) = current_user {
        let html = LoginSuccessTemplate {
            name: &user.username,
        }
        .render()
        .map_err(AuthError::from)?;

        return Ok((jar, Html(html).into_response()));
    }
    if payload.username == state.admin_username && payload.password == state.admin_password {
        // Generate JWT token
        let jwt_token = encode_jwt(payload.username.clone(), &state.jwt_secret)
            .await
            .map_err(|_| AuthError::JwtError)?;

        // Create cookie with JWT token
        let cookie = axum_extra::extract::cookie::Cookie::build(("auth_token", jwt_token))
            .http_only(true)
            .secure(false) // Set to true in production with HTTPS
            .same_site(axum_extra::extract::cookie::SameSite::Lax)
            .max_age(time::Duration::hours(24))
            .path("/")
            .build();

        let updated_jar = jar.add(cookie);

        let html = LoginSuccessTemplate {
            name: &payload.username,
        }
        .render()
        .map_err(AuthError::from)?;

        Ok((updated_jar, Html(html).into_response()))
    } else {
        let error_message = LoginErrorMessageTemplate
            .render()
            .map_err(AuthError::from)?;

        let mut headers = HeaderMap::new();
        headers.insert(
            HeaderName::from_static("hx-retarget"),
            HeaderValue::from_static("#login-message"),
        );
        headers.insert(
            HeaderName::from_static("hx-reswap"),
            HeaderValue::from_static("outerHTML"),
        );

        let mut response = Html(error_message).into_response();
        response.headers_mut().extend(headers);
        Ok((jar, response))
    }
}

pub async fn encode_jwt(username: String, jwt_secret: &str) -> anyhow::Result<String> {
    let now = chrono::Utc::now();
    let expire = chrono::Duration::hours(24);
    let exp = (now + expire).timestamp() as usize;
    let iat = now.timestamp() as usize;
    let claims = Claims { exp, iat, username };
    let jwt = encode(
        &jsonwebtoken::Header::default(),
        &claims,
        &jsonwebtoken::EncodingKey::from_secret(jwt_secret.as_bytes()),
    )?;
    Ok(jwt)
}

#[allow(dead_code)]
pub async fn decode_jwt(token: &str, jwt_secret: &str) -> anyhow::Result<Claims> {
    let token_data = jsonwebtoken::decode(
        token,
        &jsonwebtoken::DecodingKey::from_secret(jwt_secret.as_bytes()),
        &jsonwebtoken::Validation::default(),
    )?;
    Ok(token_data.claims)
}

#[derive(Template)]
#[template(path = "login_success.html")]
pub struct LoginSuccessTemplate<'a> {
    pub name: &'a str,
}

#[derive(Template)]
#[template(path = "login_error_message.html")]
pub struct LoginErrorMessageTemplate;

#[cfg(test)]
mod tests {
    use std::sync::Arc;

    use crate::config::Config;

    use super::*;
    use axum::body::Body;
    use axum::http::{Request, StatusCode};
    use tower::ServiceExt; // for `oneshot`

    async fn test_app(config: Config) -> axum::Router {
        let auth_state = Arc::new(AuthState::from_config(&config));
        super::create_login_router()
            .layer(middleware::from_fn_with_state(
                auth_state.clone(),
                super::auth_middleware,
            ))
            .with_state(auth_state)
    }

    async fn test_app_with_auth(config: Config) -> axum::Router {
        let auth_state = Arc::new(AuthState::from_config(&config));
        super::create_login_router()
            .layer(middleware::from_fn_with_state(
                auth_state.clone(),
                super::auth_middleware,
            ))
            .with_state(auth_state)
    }

    #[tokio::test]
    async fn can_login_with_valid_credentials() {
        let config = Config {
            db_url: "".to_string(),
            port: 8080,
            admin_username: "admin".to_string(),
            admin_password: "password".to_string(),
            jwt_secret: "some_secret".to_string(),
        };
        let app = test_app(config).await;

        let response = app
            .oneshot(
                Request::builder()
                    .method("POST")
                    .uri("/login")
                    .header("content-type", "application/x-www-form-urlencoded")
                    .body(Body::from("username=admin&password=password"))
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        // Check that the auth_token cookie was set
        let set_cookie_headers: Vec<_> = response.headers().get_all("set-cookie").iter().collect();
        assert!(
            !set_cookie_headers.is_empty(),
            "Expected Set-Cookie header to be present"
        );

        let cookie_header = set_cookie_headers[0].to_str().unwrap();
        assert!(
            cookie_header.contains("auth_token="),
            "Expected auth_token cookie to be set"
        );
        assert!(
            cookie_header.contains("HttpOnly"),
            "Expected HttpOnly flag to be set"
        );
        assert!(
            cookie_header.contains("Path=/"),
            "Expected Path to be set to /"
        );

        let body = axum::body::to_bytes(response.into_body(), usize::MAX)
            .await
            .unwrap();
        assert_eq!(
            body,
            LoginSuccessTemplate { name: "admin" }.render().unwrap()
        );
    }

    #[tokio::test]
    async fn can_reject_invalid_credentials() {
        let config = Config {
            db_url: "".to_string(),
            port: 8080,
            admin_username: "admin".to_string(),
            admin_password: "password".to_string(),
            jwt_secret: "some_secret".to_string(),
        };
        let app = test_app(config).await;

        let response = app
            .oneshot(
                Request::builder()
                    .method("POST")
                    .uri("/login")
                    .header("content-type", "application/x-www-form-urlencoded")
                    .body(Body::from("username=wrong&password=wrong"))
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        // Check HX-Retarget header
        let hx_retarget = response.headers().get("hx-retarget");
        assert_eq!(
            hx_retarget,
            Some(&axum::http::HeaderValue::from_static("#login-message"))
        );

        // Check HX-Reswap header
        let hx_reswap = response.headers().get("hx-reswap");
        assert_eq!(
            hx_reswap,
            Some(&axum::http::HeaderValue::from_static("outerHTML"))
        );

        let body = axum::body::to_bytes(response.into_body(), usize::MAX)
            .await
            .unwrap();
        let rendered_error = LoginErrorMessageTemplate.render().unwrap();
        assert_eq!(body, rendered_error);
        // Verify the error message is included in the response
        assert!(rendered_error.contains("Login failed. Please try again."));
    }

    #[tokio::test]
    async fn can_handle_template_error_with_internal_server_error() {
        // Simulate a template rendering error using askama::Error::Custom
        let custom_error_message = "Simulated template rendering failure".to_string();
        let template_error = askama::Error::Custom(custom_error_message.into());

        let auth_error = AuthError::Template(template_error);
        let response = axum::response::IntoResponse::into_response(auth_error);

        assert_eq!(response.status(), StatusCode::INTERNAL_SERVER_ERROR);

        let body = axum::body::to_bytes(response.into_body(), usize::MAX)
            .await
            .unwrap();
        let expected_error_message = "<h1>Internal Server Error</h1><p>An unexpected error occurred while processing your request. Please try again later.</p>";
        assert_eq!(std::str::from_utf8(&body).unwrap(), expected_error_message);
    }

    #[tokio::test]
    async fn can_return_success_when_already_logged_in() {
        let config = Config {
            db_url: "".to_string(),
            port: 8080,
            admin_username: "admin".to_string(),
            admin_password: "password".to_string(),
            jwt_secret: "some_secret".to_string(),
        };

        // First, create a valid JWT token
        let jwt_token = super::encode_jwt("admin".to_string(), &config.jwt_secret)
            .await
            .unwrap();

        let app = test_app_with_auth(config).await;

        let response = app
            .oneshot(
                Request::builder()
                    .method("POST")
                    .uri("/login")
                    .header("content-type", "application/x-www-form-urlencoded")
                    .header("cookie", format!("auth_token={}", jwt_token))
                    .body(Body::from("username=admin&password=password"))
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);

        let body = axum::body::to_bytes(response.into_body(), usize::MAX)
            .await
            .unwrap();
        assert_eq!(
            body,
            LoginSuccessTemplate { name: "admin" }.render().unwrap()
        );
    }
}
