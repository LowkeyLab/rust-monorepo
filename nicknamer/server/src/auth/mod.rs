use askama::Template;
use axum::Router;
use axum::extract::{Extension, Form, Request, State};
use axum::http::{HeaderMap, HeaderName, HeaderValue};
use axum::middleware::Next;
use axum::response::{Html, IntoResponse, Response};
use axum_extra::extract::CookieJar;
use jsonwebtoken::encode;
use std::sync::Arc;

use crate::config::Config;

/// Represents the currently authenticated user.
#[derive(Debug, Clone)]
pub struct CurrentUser {
    pub username: String,
}

impl CurrentUser {
    /// Creates a new CurrentUser instance.
    pub fn new(username: String) -> Self {
        Self { username }
    }
}

/// Authentication state containing admin credentials and JWT secret.
#[derive(Clone)]
pub struct AuthState {
    pub admin_username: String,
    pub admin_password: String,
    pub jwt_secret: String,
}

impl AuthState {
    /// Creates a new AuthState from the application config.
    pub fn from_config(config: &Config) -> Self {
        Self {
            admin_username: config.admin_username.clone(),
            admin_password: config.admin_password.clone(),
            jwt_secret: config.jwt_secret.clone(),
        }
    }
}

/// Creates a login router with authentication routes.
pub fn create_login_router(state: Arc<AuthState>) -> Router<()> {
    Router::new()
        .route("/login", axum::routing::post(login_handler))
        .route("/login", axum::routing::get(login_page_handler))
        .with_state(state)
}

/// Authentication middleware that checks for valid JWT tokens and sets CurrentUser extension.
/// This middleware only populates the CurrentUser extension and does not perform redirects.
pub async fn auth_user_middleware(
    State(state): State<Arc<AuthState>>,
    jar: CookieJar,
    mut request: Request,
    next: Next,
) -> Response {
    if let Some(token_cookie) = jar.get("auth_token") {
        if let Ok(claims) = decode_jwt(token_cookie.value(), &state.jwt_secret).await {
            let current_user = CurrentUser::new(claims.username);
            request.extensions_mut().insert(current_user);
        }
    }

    next.run(request).await
}

/// Login redirect middleware that redirects unauthenticated users to the login page.
/// This middleware should be applied after auth_user_middleware to check for CurrentUser extension.
pub async fn login_redirect_middleware(request: Request, next: Next) -> Response {
    // Check if user is authenticated by looking for CurrentUser extension
    let is_authenticated = request.extensions().get::<CurrentUser>().is_some();

    // If no valid authentication and accessing a protected route, redirect to login
    if !is_authenticated {
        return axum::response::Redirect::to("/login").into_response();
    }

    next.run(request).await
}

/// Represents the login request payload.
#[derive(serde::Deserialize, Debug)]
pub struct LoginRequest {
    pub username: String,
    pub password: String,
}

#[derive(serde::Serialize, serde::Deserialize, Debug)]
pub struct Claims {
    pub exp: usize,       // Expiry time of the token
    pub iat: usize,       // Issued at time of the token
    pub username: String, // Username of the authenticated user
}
/// Custom error type for authentication operations.
#[derive(Debug, thiserror::Error)]
pub enum AuthError {
    /// Represents an error during template rendering.
    /// The specific `askama::Error` is captured as the source of this error.
    #[error("Template rendering failed")]
    Template(#[from] askama::Error),
    /// Represents an error during JWT operations.
    /// The specific `jsonwebtoken::errors::Error` is captured as the source of this error.
    #[error("JWT operation failed")]
    JwtError,
}

impl axum::response::IntoResponse for AuthError {
    fn into_response(self) -> axum::response::Response {
        let user_facing_error_message =
            "An unexpected error occurred while processing your request. Please try again later.";
        (
            axum::http::StatusCode::INTERNAL_SERVER_ERROR,
            Html(format!(
                "<h1>Internal Server Error</h1><p>{}</p>",
                user_facing_error_message
            )),
        )
            .into_response()
    }
}

/// Handles the login request.
/// Checks submitted username and password against admin credentials.
/// If a user is already logged in, returns a success message.
pub async fn login_handler(
    State(state): State<Arc<AuthState>>,
    jar: CookieJar,
    current_user: Option<Extension<CurrentUser>>,
    Form(payload): Form<LoginRequest>,
) -> Result<(CookieJar, Response), AuthError> {
    // Check if user is already logged in
    if let Some(Extension(user)) = current_user {
        return handle_already_logged_in_user(jar, &user).await;
    }

    handle_login_attempt(state, jar, payload).await
}

/// Handles the case when a user is already logged in.
/// Returns a success response with the current user's information.
#[tracing::instrument(skip(jar))]
async fn handle_already_logged_in_user(
    jar: CookieJar,
    user: &CurrentUser,
) -> Result<(CookieJar, Response), AuthError> {
    let html = LoginSuccessTemplate {
        name: &user.username,
    }
    .render()
    .map_err(AuthError::from)?;

    Ok((jar, Html(html).into_response()))
}

/// Handles a login attempt when the user is not logged in.
/// Validates credentials and either returns success with JWT token or error response.
#[tracing::instrument(skip(state, jar, payload))]
async fn handle_login_attempt(
    state: Arc<AuthState>,
    jar: CookieJar,
    payload: LoginRequest,
) -> Result<(CookieJar, Response), AuthError> {
    if payload.username == state.admin_username && payload.password == state.admin_password {
        // Generate JWT token
        let jwt_token = encode_jwt(payload.username.clone(), &state.jwt_secret)
            .await
            .map_err(|_| AuthError::JwtError)?;

        // Create cookie with JWT token
        let cookie = axum_extra::extract::cookie::Cookie::build(("auth_token", jwt_token))
            .http_only(true)
            .secure(false) // Set to true in production with HTTPS
            .same_site(axum_extra::extract::cookie::SameSite::Lax)
            .max_age(time::Duration::hours(24))
            .path("/")
            .build();

        let updated_jar = jar.add(cookie);

        let html = LoginSuccessTemplate {
            name: &payload.username,
        }
        .render()
        .map_err(AuthError::from)?;

        Ok((updated_jar, Html(html).into_response()))
    } else {
        let error_message = LoginErrorMessageTemplate
            .render()
            .map_err(AuthError::from)?;

        let mut headers = HeaderMap::new();
        headers.insert(
            HeaderName::from_static("hx-retarget"),
            HeaderValue::from_static("#login-message"),
        );
        headers.insert(
            HeaderName::from_static("hx-reswap"),
            HeaderValue::from_static("outerHTML"),
        );

        let mut response = Html(error_message).into_response();
        response.headers_mut().extend(headers);
        Ok((jar, response))
    }
}

pub async fn encode_jwt(username: String, jwt_secret: &str) -> anyhow::Result<String> {
    let now = chrono::Utc::now();
    let expire = chrono::Duration::hours(24);
    let exp = (now + expire).timestamp() as usize;
    let iat = now.timestamp() as usize;
    let claims = Claims { exp, iat, username };
    let jwt = encode(
        &jsonwebtoken::Header::default(),
        &claims,
        &jsonwebtoken::EncodingKey::from_secret(jwt_secret.as_bytes()),
    )?;
    Ok(jwt)
}

#[allow(dead_code)]
pub async fn decode_jwt(token: &str, jwt_secret: &str) -> anyhow::Result<Claims> {
    let token_data = jsonwebtoken::decode(
        token,
        &jsonwebtoken::DecodingKey::from_secret(jwt_secret.as_bytes()),
        &jsonwebtoken::Validation::default(),
    )?;
    Ok(token_data.claims)
}

#[derive(Template)]
#[template(path = "login/login_success.html")]
pub struct LoginSuccessTemplate<'a> {
    pub name: &'a str,
}

#[derive(Template)]
#[template(path = "login/login_error_message.html")]
pub struct LoginErrorMessageTemplate;

#[derive(Template)]
#[template(path = "login.html")]
pub struct LoginTemplate {
    pub username: Option<String>,
}

/// Handles GET requests to display the login page.
#[tracing::instrument]
pub async fn login_page_handler(
    current_user: Option<Extension<CurrentUser>>,
) -> Result<Html<String>, AuthError> {
    let username = current_user.map(|Extension(user)| user.username);

    let template = LoginTemplate { username };
    template.render().map(Html).map_err(AuthError::from)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::config::Config;

    #[tokio::test]
    async fn auth_middlewares_work_together() {
        use axum::body::Body;
        use axum::http::{Request, StatusCode};
        use axum::middleware::from_fn_with_state;
        use tower::ServiceExt;

        let config = Config {
            db_url: "".to_string(),
            port: 8080,
            admin_username: "admin".to_string(),
            admin_password: "password".to_string(),
            jwt_secret: "test_secret".to_string(),
        };

        let auth_state = Arc::new(AuthState::from_config(&config));

        // Create a test app with both middlewares in the correct order
        // Note: Layers are applied in reverse order (bottom to top)
        let app = axum::Router::new()
            .route(
                "/protected",
                axum::routing::get(|| async { "Protected content" }),
            )
            .layer(axum::middleware::from_fn(login_redirect_middleware))
            .layer(from_fn_with_state(auth_state.clone(), auth_user_middleware));

        // Test 1: Unauthenticated request should redirect to login
        let response = app
            .clone()
            .oneshot(
                Request::builder()
                    .method("GET")
                    .uri("/protected")
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::SEE_OTHER);
        let location = response.headers().get("location").unwrap();
        assert_eq!(location, "/login");

        // Test 2: Authenticated request should allow access
        let jwt_token = encode_jwt("admin".to_string(), &config.jwt_secret)
            .await
            .unwrap();

        let response = app
            .oneshot(
                Request::builder()
                    .method("GET")
                    .uri("/protected")
                    .header("cookie", format!("auth_token={}", jwt_token))
                    .body(Body::empty())
                    .unwrap(),
            )
            .await
            .unwrap();

        assert_eq!(response.status(), StatusCode::OK);
        let body = axum::body::to_bytes(response.into_body(), usize::MAX)
            .await
            .unwrap();
        assert_eq!(body, "Protected content");
    }
}

pub mod api;
