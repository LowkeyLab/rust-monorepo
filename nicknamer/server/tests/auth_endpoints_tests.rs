use axum::body::Body;
use axum::extract::Extension;
use axum::http::Request;
use axum::middleware::{from_fn, from_fn_with_state};
use insta::assert_yaml_snapshot;
use nicknamer_server::auth::{
    AuthError, AuthState, CurrentUser, create_login_router, encode_jwt, login_page_handler,
};
use nicknamer_server::config::Config;
use std::sync::Arc;
use tower::ServiceExt;

mod common;

use common::{HttpResponseSnapshot, create_stub_user_middleware};

/// Setup function for auth endpoint tests.
async fn setup_auth_state() -> Arc<AuthState> {
    let config = Config {
        db_url: "".to_string(),
        port: 8080,
        admin_username: "admin".to_string(),
        admin_password: "password".to_string(),
        jwt_secret: "some_secret".to_string(),
    };
    Arc::new(AuthState::from_config(&config))
}

/// Test helper to create test app with auth state.
async fn create_test_app() -> (axum::Router, Arc<AuthState>) {
    let auth_state = setup_auth_state().await;
    let app = create_login_router(auth_state.clone()).layer(from_fn_with_state(
        auth_state.clone(),
        nicknamer_server::auth::auth_user_middleware,
    ));
    (app, auth_state)
}

/// Test helper to create test app with a logged-in user.
async fn create_test_app_with_logged_in_user(username: String) -> (axum::Router, Arc<AuthState>) {
    let auth_state = setup_auth_state().await;
    let app = create_login_router(auth_state.clone())
        .layer(from_fn(create_stub_user_middleware(username)));
    (app, auth_state)
}

#[tokio::test]
async fn can_login_with_valid_credentials() {
    let (app, _auth_state) = create_test_app().await;

    let request = Request::builder()
        .method("POST")
        .uri("/login")
        .header("content-type", "application/x-www-form-urlencoded")
        .body(Body::from("username=admin&password=password"))
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    let status = response.status();
    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data =
        HttpResponseSnapshot::new(body_text, status, &headers, "login_with_valid_credentials");

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_reject_invalid_credentials() {
    let (app, _auth_state) = create_test_app().await;

    let request = Request::builder()
        .method("POST")
        .uri("/login")
        .header("content-type", "application/x-www-form-urlencoded")
        .body(Body::from("username=wrong&password=wrong"))
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    let status = response.status();
    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data =
        HttpResponseSnapshot::new(body_text, status, &headers, "reject_invalid_credentials");

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_handle_template_error_with_internal_server_error() {
    // Simulate a template rendering error using askama::Error::Custom
    let custom_error_message = "Simulated template rendering failure".to_string();
    let template_error = askama::Error::Custom(custom_error_message.into());

    let auth_error = AuthError::Template(template_error);
    let response = axum::response::IntoResponse::into_response(auth_error);

    let status = response.status();
    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data =
        HttpResponseSnapshot::new(body_text, status, &headers, "handle_template_error");

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_return_success_when_already_logged_in() {
    let (app, auth_state) = create_test_app().await;

    // First, create a valid JWT token
    let jwt_token = encode_jwt("admin".to_string(), &auth_state.jwt_secret)
        .await
        .unwrap();

    let request = Request::builder()
        .method("POST")
        .uri("/login")
        .header("content-type", "application/x-www-form-urlencoded")
        .header("cookie", format!("auth_token={}", jwt_token))
        .body(Body::from("username=admin&password=password"))
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    let status = response.status();
    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        status,
        &headers,
        "return_success_when_already_logged_in",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_display_login_page() {
    let (app, _auth_state) = create_test_app().await;

    let request = Request::builder()
        .method("GET")
        .uri("/login")
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    let status = response.status();
    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data =
        HttpResponseSnapshot::new(body_text, status, &headers, "display_login_page");

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_display_login_page_with_homepage_button_when_logged_in() {
    let (app, _auth_state) = create_test_app_with_logged_in_user("admin".to_string()).await;

    let request = Request::builder()
        .method("GET")
        .uri("/login")
        .body(Body::empty())
        .unwrap();

    let response = app.oneshot(request).await.unwrap();

    let status = response.status();
    let headers = response.headers().clone();
    let body = axum::body::to_bytes(response.into_body(), usize::MAX)
        .await
        .unwrap();
    let body_text = std::str::from_utf8(&body).unwrap();

    let snapshot_data = HttpResponseSnapshot::new(
        body_text,
        status,
        &headers,
        "display_login_page_with_homepage_button_when_logged_in",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_render_login_page_form_when_user_not_logged_in() {
    let result = login_page_handler(None).await;

    assert!(result.is_ok());
    let html = result.unwrap().0;

    let snapshot_data = HttpResponseSnapshot::new(
        &html,
        axum::http::StatusCode::OK,
        &axum::http::HeaderMap::new(),
        "render_login_page_form_when_user_not_logged_in",
    );

    assert_yaml_snapshot!(snapshot_data);
}

#[tokio::test]
async fn can_render_login_page_with_homepage_button_when_user_logged_in() {
    let current_user = CurrentUser::new("testuser".to_string());
    let extension = Extension(current_user);

    let result = login_page_handler(Some(extension)).await;

    assert!(result.is_ok());
    let html = result.unwrap().0;

    let snapshot_data = HttpResponseSnapshot::new(
        &html,
        axum::http::StatusCode::OK,
        &axum::http::HeaderMap::new(),
        "render_login_page_with_homepage_button_when_user_logged_in",
    );

    assert_yaml_snapshot!(snapshot_data);
}
