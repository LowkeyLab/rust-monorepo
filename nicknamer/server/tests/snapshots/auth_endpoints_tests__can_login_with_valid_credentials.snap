---
source: nicknamer/server/tests/auth_endpoints_tests.rs
expression: snapshot_data
---
test_context: login_with_valid_credentials
status: 200
headers:
  content-type: text/html; charset=utf-8
html_body:
  - "<div class=\"card-body items-center text-center\">"
  - "  <h2 class=\"card-title\">Login Successful!</h2>"
  - "  <p>Hello <b>admin</b></p>"
  - "  <div hx-boost=\"true\" class=\"card-actions justify-end\">"
  - "    <a href=\"/names\" class=\"btn btn-primary\">Manage Names</a>"
  - "  </div>"
  - "</div>"
